src/session-pool.ts

```diff
@@ -752,41 +758,54 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
     let nReturned = 0;
     const nRequested: number = amount;
 
-    // while we can request as many sessions be created as we want, the backend
-    // will return at most 100 at a time, hence the need for a while loop.
-    while (amount > 0) {
-      let sessions: Session[] | null = null;
+    if (!this.database._observabilityOptions) {
+      this.database._observabilityOptions = this._observabilityOptions;
+    }
 
-      span.addEvent(`Creating ${amount} sessions`);
+    const q = {opts: this._observabilityOptions};
+    return startTrace('SessionPool.createSessions', q, async span => {
+      span.addEvent(`Requesting ${amount} sessions`);
+
+      // while we can request as many sessions be created as we want, the backend
+      // will return at most 100 at a time, hence the need for a while loop.
+      while (amount > 0) {
+        let sessions: Session[] | null = null;
+
+        span.addEvent(`Creating ${amount} sessions`);
+
+        try {
+          [sessions] = await this.database.batchCreateSessions({
+            count: amount,
+            labels: labels,
+            databaseRole: databaseRole,
+          });
+
+          amount -= sessions.length;
+          nReturned += sessions.length;
+        } catch (e) {
+          this._pending -= amount;
+          this.emit('createError', e);
+          span.addEvent(
+            `Requested for ${nRequested} sessions returned ${nReturned}`
+          );
+          setSpanErrorAndException(span, e as Error);
+          throw e;
```

Comment:
There should be a `span.end()` in the catch block ? before throwing error ? 
If this is true , also add a test case which could have caught this

Reply:
Good catch and thank you! I've added a check and a fix too.